### PR TITLE
feat: regex filters for namespace and name

### LINKED ISSUE #582: Look Into Allowing Regex in InNamespace
### Is your feature request related to a problem? Please describe.
For uds-core policy exemptions, there might be a case for optionally locking down which namespace the operator watches for Exemption CRs. With the current implementation of `.InNamespace()`, it does not look to be possible to set it to a single namespace or list / all namespaces based on env var. (e.g. `.InNamespace(process.env.EXEMPTION_NS)`)

### Describe the solution you'd like

- Given an admin wants the option to allow single or all namespaces in `.InNamespace()`, or maybe namespaces that fit a pattern
- he or she could use a regex patterns like so `.InNamspace(".*")` , `.InNamespace("uds-policy-exemptions")`, `.InNamespace("*-policy-exemptions")`

### Describe alternatives you've considered
The alternative is to do this as the module author in the validator and not use `.InNamespace()`.

### PR DESCRIPTION
**Pull request by cmwylie19**
## Description

Adds regex for name and namespaces. 

Another note: Filter Logic is extremely difficult to grok through now, we need to [refactor](https://github.com/defenseunicorns/pepr/issues/1154)

**TODO**
- [x] e2e test
- [x] Extend Unit tests to add cases for regex
- [x] `namespaceComplianceValidator` to test for regex namespaces

End to End Test: [PR](https://github.com/defenseunicorns/pepr-excellent-examples/pull/75)
- [x] hello-pepr-regex-name (should fail until #1162 lands)
- [x] hello-pepr-regex-ns-all
- [x] hello-pepr-regex-ns-bounded
- [x] hello-pepr-regex-ns-wrong

## Related Issue

Fixes #582 

Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging
- [x] Unit, [Journey](https://github.com/defenseunicorns/pepr/tree/main/journey), [E2E Tests](https://github.com/defenseunicorns/pepr-excellent-examples), [docs](https://github.com/defenseunicorns/pepr/tree/main/docs), [adr](https://github.com/defenseunicorns/pepr/tree/main/adr) added or updated as needed
- [x] [Contributor Guide Steps](https://docs.pepr.dev/main/contribute/#submitting-a-pull-request) followed

### CODE REVIEW

File: src/lib/helpers.ts [hunk 1]
```diff
@@ -7,6 +7,20 @@ import Log from "./logger";
...
+  if (!pattern) {
+    return false;
+  }
+
+  try {
+    const regex = new RegExp(pattern);
+    return regex.test(testString);
+  } catch {
+    return false;
+  }
```
Comment (cmwylie19) [inline src/lib/helpers.ts]:
```suggestion
  const regex = new RegExp(pattern);
  return regex.test(testString);
```

Pretty sure this try catch is unnecessary and brings down coverage due to the fact that you can only add Regex to `WithRegexName` and `InRegexNamespace`, the error will never get to this function.

File: docs/030_user-guide/130_filters.md [hunk 2]
```diff
@@ -24,7 +24,9 @@ When(a.ConfigMap)
 ## `Filters`
 
 - `.WithName("name")`: Filters resources by name.
+- `.WithRegexName(/^pepr/)`: Filters resources by name using a regex.
 - `.InNamespace("namespace")`: Filters resources by namespace.
+- `.WithRegexName(/(.*)-system/)`: Filters resources by name using a regex.
```
Comment (cmwylie19) [inline docs/030_user-guide/130_filters.md]:
```suggestion
- `.WithRegexNamespace(/(.*)-system/)`: Filters resources by namespace using a regex.
```

File: docs/030_user-guide/130_filters.md [hunk 3]
```diff
@@ -24,7 +24,9 @@ When(a.ConfigMap)
 ## `Filters`
 
 - `.WithName("name")`: Filters resources by name.
+- `.WithRegexName(/^pepr/)`: Filters resources by name using a regex.
 - `.InNamespace("namespace")`: Filters resources by namespace.
+- `.WithRegexNamespace(/(.*)-system/)`: Filters resources by namespace using a regex.
```
Comment (cmwylie19) [inline docs/030_user-guide/130_filters.md]:
```suggestion
- `.InRegexNamespace(/(.*)-system/)`: Filters resources by namespace using a regex.
```

File: src/lib/capability.ts [hunk 4]
```diff
@@ -321,6 +329,12 @@ export class Capability implements CapabilityExport {
       return commonChain;
     }
 
+    function WithRegexName(regexName: RegExp): BindingFilter<T> {
```
Comment (cmwylie19) [inline src/lib/capability.ts]: WithNameRegex

File: src/lib/filter.test.ts [hunk 5]
```diff
@@ -105,25 +105,182 @@ describe("Property-Based Testing shouldSkipRequest", () => {
...
+test("create: should not reject when regex namespace does match", () => {
+  const binding = {
+    model: kind.Pod,
+    event: Event.Any,
+    kind: podKind,
+    filters: {
+      name: "",
+      namespaces: [],
+      regexNamespaces: [new RegExp(/^helm/).source],
+      regexName: new RegExp("").source,
```
Comment (cmwylie19) [inline src/lib/filter.test.ts]:
```suggestion
      regexName: "",
```

File: src/lib/filter.test.ts [hunk 6]
```diff
@@ -105,25 +105,182 @@ describe("Property-Based Testing shouldSkipRequest", () => {
...
+test("create: should reject when regex namespace does not match", () => {
+  const binding = {
+    model: kind.Pod,
+    event: Event.Any,
+    kind: podKind,
+    filters: {
+      name: "",
+      namespaces: [],
+      regexNamespaces: [new RegExp(/^argo/).source],
+      regexName: new RegExp("").source,
```
Comment (cmwylie19) [inline src/lib/filter.test.ts]:
```suggestion
      regexName: "",
```

**Review by btlghrants — APPROVED** [review]
Given our (lengthy) in-person review earlier, your need to move over to an alternate-watcher implementation post-haste, and the fact that we already discussed how a follow-on refactor story will be coming out of the this + what I'm working on currently, this PR is acceptable.

**Pull request merged**